text/0080-datetime-extension.md

```diff
@@ -0,0 +1,365 @@
+It is also difficult for the calling application to predict the necessary pre-computed values that a policy writer requires for their intended purpose. These may change over time, and may also differ depending on the principal, action, and/or resource.
+
+### Additional examples
+
+**Only allow user "alice" to view JPEG photos for one week after creation.**
+
+```cedar
+permit(
+  principal == User::"alice",
+  action == PhotoOp::"view",
+  resource is Photo
+) when {
+  resource.fileType == "JPEG" &&
+  (context.now.timestamp - resource.creationTime) <= duration(7, "days")
+};
+```
+
+**Allow access from a certain IP address only between 9am and 6pm UTC.**
+
+```cedar
+permit(
+  principal,
+  action == Action::"access",
+  resource
+) when {
+  context.srcIp.isInRange(ip("192.168.1.0/24")) &&
+  context.workdayStart <= context.now.timestamp &&
+  context.now.timestamp < context.workdayEnd
+};
+```
+
+**Allow access from a certain IP address only between 9am and 6pm localized to their time.**
+
+```cedar
+permit(
+  principal,
+  action == Action::"access",
+  resource
+) when {
+  context.srcIp.isInRange(ip("192.168.1.0/24")) &&
+  context.workdayStart <= context.now.timestamp &&
+  context.now.timestamp < context.workdayEnd
+};
+```
+
+Note that this is identical to the previous policy. The calling application must adjust `workdayStart` and `workdayEnd`.
+
+**Prevent employees from accessing work documents on the weekend.**
+
+```cedar
+forbid(
+  principal,
+  action == Action::"access",
+  resource is Document
+) when {
+  [1,7].contains(context.now.dayOfWeek)
+};
+```
+
+**Permit access to a special opportunity for persons born on Leap Day.**
+
+```cedar
+permit(
+  principal,
+  action == Action::"redeem",
+  resource is Prize
+) when {
+  principal.birthDate.day == 29 &&
+  principal.birthDate.month == 2 &&
+  context.now.day == 29 && 
+  context.now.month == 2
+};
+```
+
+**Forbid access to EU resources after Brexit**
+
+```cedar
+forbid(
+  principal,
+  action,
+  resource
+) when {
+  context.now.timestamp > datetime("20200131T23:00:00Z") &&
+  context.location.countryOfOrigin == 'GB' &&
+  resource.owner == 'EU'
+}
+```
+
+## Detailed design
+
+This RFC proposes supporting two new extension types: `datetime`, which represents a particular instant of time, up to millisecond accuracy, and `duration` which represents a duration of time.
+To construct and manipulate these types we will provide the functions listed below.
+All of this functionality will be hidden behind a `datetime` feature flag (analogous to the current decimal and IP extensions), allowing users to opt-out.
+
+- `datetime(string)` constructs a datetime value. Like with other extension function constructors, strict validation requires `string` to be a string literal, although evaluation/authorization support any string-typed expression. The string must be of one of the forms, and regardless of the timezone offset is always normalized to UTC:
```

Comment:
Maybe we should also include `datetime("epoch_start")` and `datetime("epoch_end")`?
A potential usecase would be half-open intervals. E.g., the same shape of policy could be used to express buying and renting. For buying it could be:
```
permit(principal = User::"Bob", action = Action::"view", resource = Movie::"UnReleasedFilm") when {
    context.curr_time >= resource.release_date && context.curr_time <= datetime("epoch_end")
};
```
While for renting you'd use the end of the rental period, e.g., `context.curr_time <= datetime("2024-08-21T20:07:23")`.

Reply:
This seems fine to me. Perhaps this is mostly useful when policies are templated, possibly in the context of #61, but why would a policy author include `context.curr_time <= datetime("epoch_end")` otherwise?

Reply:
Yeah, I'm envisioning templated / programmatically generated policies where it's easier to just give every policy a start and end time. In my example the RHS condition would always true so it would hopefully be dropped if the policies were written by hand.

Reply:
I don't think I understand the utility here: if you want to time-box `context.curr_time`, wouldn't it make more sense to give specific values for the bounds rather than using `epoch_start` and/or `epoch_end`? 

I'm not sure that `epoch_start` and `epoch_end` are useful abstractions here: one is in 1970, and the other is the largest representable 64-bit value, the date of which will vary with our supported precision level.

Reply:
I think the usefulness is only in generated policies, but also agree with @khieta that the user could provide a date far into the future, or far into the past just as easily. I'm :-1: on adding `epoch_start` and `epoch_end`.